internal/pkg/service/stream/storage/node/readernode/sliceupload/operator.go

```diff
@@ -0,0 +1,363 @@
+			d.StorageRepository().Slice().GetAllInLevelAndWatch(ctx, model.LevelLocal, etcd.WithPrevKV()),
+			func(_ string, slice model.Slice) model.SliceKey {
+				return slice.SliceKey
+			},
+			func(_ string, slice model.Slice, rawValue *op.KeyValue, oldValue **sliceData) *sliceData {
+				out := &sliceData{
+					Slice: &slice,
+				}
+
+				// Keep the same lock, to prevent parallel processing of the same slice.
+				// No modification from another code is expected, but just to be sure.
+				if oldValue != nil {
+					out.Lock = (*oldValue).Lock
+				} else {
+					out.Lock = &sync.Mutex{}
+				}
+
+				return out
+			},
+		).
+			WithFilter(func(event etcdop.WatchEvent[model.Slice]) bool {
+				return o.volumes.Collection().HasVolume(event.Value.VolumeID)
+			}).
+			BuildMirror()
+		if err = <-o.slices.StartMirroring(ctx, wg, o.logger); err != nil {
+			return err
+		}
+	}
+
+	// Mirror already uploaded slices for updating manifest
+	{
+		o.uploadedSlices = etcdop.SetupMirrorMap[model.Slice, model.SliceKey, *uploadedSliceData](
+			d.StorageRepository().Slice().GetAllInLevelAndWatch(ctx, model.LevelStaging, etcd.WithPrevKV()),
+			func(_ string, slice model.Slice) model.SliceKey {
+				return slice.SliceKey
+			},
+			func(_ string, slice model.Slice, rawValue *op.KeyValue, oldValue **uploadedSliceData) *uploadedSliceData {
+				out := &uploadedSliceData{
+					filename: slice.String(),
+				}
+
+				// Keep the same lock, to prevent parallel processing of the same slice.
+				// No modification from another code is expected, but just to be sure.
+				if oldValue != nil {
+					out.Lock = (*oldValue).Lock
+				} else {
+					out.Lock = &sync.Mutex{}
+				}
+
+				return out
+			},
+		).
+			WithFilter(func(event etcdop.WatchEvent[model.Slice]) bool {
+				return event.Value.State == model.SliceUploaded
+			}).
+			BuildMirror()
+		if err = <-o.uploadedSlices.StartMirroring(ctx, wg, o.logger); err != nil {
+			return err
+		}
+	}
+	// Start conditions check ticker
+	{
+		wg.Add(1)
+		ticker := d.Clock().Ticker(o.config.CheckInterval.Duration())
+
+		go func() {
+			defer wg.Done()
+			defer ticker.Stop()
+
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-ticker.C:
+					o.checkSlices(ctx, wg)
+				}
+			}
+		}()
+	}
+
+	return nil
+}
+
+func (o *operator) checkSlices(ctx context.Context, wg *sync.WaitGroup) {
+	o.logger.Debugf(ctx, "checking slices upload conditions")
+
+	o.slices.ForEach(func(_ model.SliceKey, data *sliceData) (stop bool) {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			o.checkSlice(ctx, data)
+		}()
+		return false
+	})
+}
+
+func (o *operator) checkSlice(ctx context.Context, data *sliceData) {
+	// Prevent multiple checks of the same slice
+	if !data.Lock.TryLock() {
+		return
+	}
+	defer data.Lock.Unlock()
+
+	// No plugin existing for given slice to be uploaded
+	err := o.plugins.ImporterFor(data.Slice.StagingStorage.Provider)
```

Comment:
Check that we have plugin available for slice `StagingStorage.Provider`

Reply:
Typo? `ImporterFor` in `slice upload` operation.

Reply:
Removed. Check done in `UploadSlice`.